### PR TITLE
[MAPNIK] new port added

### PR DESCRIPTION
**Pull request by am2222**
**Describe the pull request**

- What does your PR fix? Fixes #
Adds new port for mapnik https://github.com/microsoft/vcpkg/issues/7526
- Which triplets are supported/not supported? Have you updated the CI baseline?
```
mapnik:arm-uwp=fail
mapnik:x64-uwp=fail
 ```
- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
Yes

This port has a dependency to the following package
https://github.com/microsoft/vcpkg/pull/11652

### CODE REVIEW

File: scripts/ci.baseline.txt [hunk 1]
```diff
@@ -2033,3 +2033,6 @@ ctp:x64-uwp=skip
 ctp:x64-windows=skip
 ctp:x64-windows-static=skip
 ctp:x86-windows=skip
+
+mapnik:arm-uwp=fail
+mapnik:x64-uwp=fail
```
Comment (NancyLi1013) [inline scripts/ci.baseline.txt:2038]: Please put these two lines at the place that starts with `m` and keep the codes alphabetically.

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

File: ports/mapnik/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,210 @@
+set(MAPNIK_VERSION 3.0.22)
+include(vcpkg_common_functions)
+find_program(GIT git)
+
+set(GIT_URL "https://github.com/am2222/mapnik-windows.git")
+set(GIT_REV "fdba45cd95b66576e47de2e56d30196bfc9de99d")
+
+set(SOURCE_PATH ${CURRENT_BUILDTREES_DIR}/src/${PORT})
```
[the remark this answers is not in the document]
Comment (am2222) [reply ports/mapnik/portfile.cmake:8]: I had to pull the repository with `--recurse-submodules` which I think `vcpkg+from_github()` does not support it.

File: ports/mapnik/portfile.cmake [hunk 3]
```diff
@@ -0,0 +1,210 @@
+set(MAPNIK_VERSION 3.0.22)
+include(vcpkg_common_functions)
```
[the remark this answers is not in the document]
Comment (am2222) [reply ports/mapnik/portfile.cmake:2]: Done!

File: ports/mapnik/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+	  WORKING_DIRECTORY ${SOURCE_PATH}
+	  LOGNAME checkout
+	)
+endif()
+message(STATUS "Adding worktree done")
+
+
+vcpkg_find_acquire_program(PYTHON2)
+get_filename_component(PYTHON2_DIR "${PYTHON2}" DIRECTORY)
+set(ENV{PATH} "$ENV{PATH};${PYTHON2_DIR}")
```
[the remark this answers is not in the document]
Comment (am2222) [reply ports/mapnik/portfile.cmake:30]: Done!

File: ports/mapnik/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+if(DEBUG_EXES)
+    file(REMOVE ${EXES})
+endif()
+
+
+
+
+#file(REMOVE_RECURSE  "${CURRENT_PACKAGES_DIR}/debug")
+file(INSTALL ${SOURCE_PATH}/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/mapnik RENAME copyright)
+file(INSTALL ${SOURCE_PATH}/fonts/unifont_license.txt DESTINATION ${CURRENT_PACKAGES_DIR}/share/mapnik RENAME fonts_copyright)
```
[the remark this answers is not in the document]
Comment (am2222) [reply ports/mapnik/portfile.cmake:210]: Done!

File: scripts/ci.baseline.txt (hunk 1 above)
Comment (am2222) [reply scripts/ci.baseline.txt:2038]: Done!

File: ports/mapnik/portfile.cmake (hunk 2 above)
[the remark this answers is not in the document]
Comment (NancyLi1013) [reply ports/mapnik/portfile.cmake:8]: Why do you need do this?

**Comment by NancyLi1013** [comment]
It seems the review suggestions are not been resolved yet.
Could you please check them again?

**Comment by am2222** [comment]
> It seems the review suggestions are not been resolved yet.
> Could you please check them again?

Hi, I had addressed them but I was trying to figure out the reason why it fails on `x64-windows` and `Linux` but I found out this error for Linux build.
https://github.com/microsoft/vcpkg/issues/11780
and on my machine `x64-windows` builds with no issues but I think here it fails.

File: ports/mapnik/CONTROL [hunk 6]
```diff
@@ -0,0 +1,58 @@
+Source: mapnik
+Version: 3.0.22
+Homepage: https://github.com/am2222/mapnik-windows/
+Description: Mapnik is an open source toolkit for developing mapping applications. Compiling on windows. (unofficial port)
+Build-Depends: icu, boost-locale[icu], boost-regex[icu], boost, freetype, tiff, libwebp, libxml2, harfbuzz, libpng, zlib
+
+Feature: WITH_CAIRO
```
[the remark this answers is not in the document]
Comment (NancyLi1013) [reply ports/mapnik/CONTROL]: The feature name is not updated.

**Comment by NancyLi1013** [comment]
OK. You can push them after you solved the failures. I will try to review this PR again when you have done all work for this PR.

**Comment by am2222** [comment]
Thanks. I pushed almost all of the changes. however still waiting for the results of the checks for x64-windows and linux.

**Comment by NancyLi1013** [comment]
@am2222 
Could you please resolve the conflicts?

**Comment by am2222** [comment]
@NancyLi1013 Thanks. I did not notice that conflict. I resolved it.

File: ports/mapnik/portfile.cmake (hunk 2 above)
[the remark this answers is not in the document]
Comment (am2222) [reply ports/mapnik/portfile.cmake]: I think now that I am fetching data from mapnik's repo this issue is also resolved.

File: ports/mapnik/CONTROL [hunk 7]
```diff
@@ -0,0 +1,58 @@
+Source: mapnik
+Version: 3.0.22
+Homepage: https://github.com/mapnik/mapnik
+Description: Mapnik is an open source toolkit for developing mapping applications. 
+Build-Depends: libjpeg-turbo, tiff, libpng, libwebp, freetype[bzip2,png], icu, libxml2, harfbuzz[icu],  boost-locale[icu], boost-regex[icu],  boost-gil, boost-geometry, boost-property-tree, boost-filesystem, boost-system, boost-spirit, boost-program-options, boost-assign, boost-bimap, mapbox-variant, zlib
+
+Feature: CAIRO
```
Comment (NancyLi1013) [inline ports/mapnik/CONTROL:7]:
```suggestion
Feature: cairo
```
Could you please update the feature name? We prefer to use lowercase here.

File: ports/mapnik/CONTROL [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+Version: 3.0.22
+Homepage: https://github.com/mapnik/mapnik
+Description: Mapnik is an open source toolkit for developing mapping applications. 
+Build-Depends: libjpeg-turbo, tiff, libpng, libwebp, freetype[bzip2,png], icu, libxml2, harfbuzz[icu],  boost-locale[icu], boost-regex[icu],  boost-gil, boost-geometry, boost-property-tree, boost-filesystem, boost-system, boost-spirit, boost-program-options, boost-assign, boost-bimap, mapbox-variant, zlib
+
+Feature: CAIRO
+Description: Cairo renderer
+Build-Depends: cairo,cairomm
+
+Feature: PROJ4
```
Comment (NancyLi1013) [inline ports/mapnik/CONTROL:11]: The same as above.

File: ports/mapnik/portfile.cmake [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+# 	  LOGNAME clone
+# 	)
+
+# 	message(STATUS "Checkout revision ${GIT_REV}")
+# 	vcpkg_execute_required_process(
+# 	  COMMAND ${GIT} checkout ${GIT_REV}
+# 	  WORKING_DIRECTORY ${SOURCE_PATH}
+# 	  LOGNAME checkout
+# 	)
+# endif()
```
Comment (NancyLi1013) [inline ports/mapnik/portfile.cmake:21]: Please remove these comments codes here if they are no longer needed.

File: ports/mapnik/portfile.cmake [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO mapnik/mapnik
+    REF v3.0.23
+    SHA512 004429defd3047601b63c28bedd57d701ff88cdfb910a9958f4a10a1ac032d4d13316372122aef06eda60f2e4f67600b8842577db455f8c970009a0e86c5c1cf
+    HEAD_REF master
+    PATCHES 0001-CMakeFiles.patch 0002-sqlitepixelwidth.patch
```
Comment (NancyLi1013) [inline ports/mapnik/portfile.cmake:29]:
```suggestion
    PATCHES
        0001-CMakeFiles.patch 
        0002-sqlitepixelwidth.patch
```
Could you please update this?

File: ports/mapnik/portfile.cmake [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+)
+
+message(STATUS "Adding worktree done")
```
Comment (NancyLi1013) [inline ports/mapnik/portfile.cmake:32]:
```suggestion
```
This line seems to be unnecessary.

File: ports/mapnik/portfile.cmake [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+vcpkg_add_to_path(${PYTHON2_DIR})
+
+set(BOOST_ROOT ${CURRENT_INSTALLED_DIR})
+
+if (VCPKG_LIBRARY_LINKAGE STREQUAL dynamic)
+  set(MAPNIK_STATIC_LIB OFF)
+elseif (VCPKG_LIBRARY_LINKAGE STREQUAL static)
+  set(MAPNIK_STATIC_LIB ON)
+endif()
+
```
Comment (NancyLi1013) [inline ports/mapnik/portfile.cmake:46]:
```suggestion
```
Please remove this blank line.

File: ports/mapnik/portfile.cmake [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+  # Keyword FEATURES is optional if INVERTED_FEATURES are not used
```
Comment (NancyLi1013) [inline ports/mapnik/portfile.cmake:49]:
```suggestion
```
This comment should be removed here.

File: ports/mapnik/portfile.cmake [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    INPUT_GEOJSON   WITH_INPUT_GEOJSON
+    INPUT_OGR WITH_INPUT_OGR
+    INPUT_PGRASTER  WITH_INPUT_PGRASTER
+    INPUT_POSTGIS  WITH_INPUT_POSTGIS
+    INPUT_RASTER  WITH_INPUT_RASTER
+    INPUT_SHAPE WITH_INPUT_SHAPE
+    INPUT_SQLITE WITH_INPUT_SQLITE
+    INPUT_TOPOJSON  WITH_INPUT_TOPOJSON
+    PROJ4 WITH_PROJ4
+    UTILS WITH_UTILS
```
Comment (NancyLi1013) [inline ports/mapnik/portfile.cmake:64]: Could you please update the feature names as lowercase and also in CONTROL file?

File: ports/mapnik/portfile.cmake [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+
+)
+
+
```
Comment (NancyLi1013) [inline ports/mapnik/portfile.cmake:68]:
```suggestion
```
Please remove this blank line.

File: ports/mapnik/portfile.cmake [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+vcpkg_configure_cmake(
+    SOURCE_PATH ${CURRENT_BUILDTREES_DIR}/src/${PORT}
+ #    PREFER_NINJA # Disable this option if project cannot be built with Ninja
```
Comment (NancyLi1013) [inline ports/mapnik/portfile.cmake:71]:
```suggestion
       PREFER_NINJA # Disable this option if project cannot be built with Ninja
```
Why do you comment this line out? Can this port not be built with Ninja? If so, please remove this line.

File: ports/mapnik/portfile.cmake [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+#    NO_CHARSET_FLAG # automatic templates
```
Comment (NancyLi1013) [inline ports/mapnik/portfile.cmake:72]:
```suggestion
```
This line seems to be not needed. Please remove it here.

File: ports/mapnik/portfile.cmake [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+      ${FEATURE_OPTIONS}
+    	-DMAPNIK_STATIC_LIB=${MAPNIK_STATIC_LIB}
+    	-DBOOST_PREFIX=${CURRENT_INSTALLED_DIR}/include
+    	-DFREE_TYPE_INCLUDE=${CURRENT_INSTALLED_DIR}/include/freetype2
+)
+
+vcpkg_install_cmake()
+
+
+
```
Comment (NancyLi1013) [inline ports/mapnik/portfile.cmake:83]: Please remove the extra blank lines.

File: ports/mapnik/portfile.cmake [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+    vcpkg_copy_tools(
+      TOOL_NAMES mapnik-demo mapnik-render shapeindex
+      AUTO_CLEAN
+    )
+    file(COPY ${SOURCE_PATH}/demo/data DESTINATION ${CURRENT_PACKAGES_DIR}/tools/mapnik/demo)
+    file(COPY ${SOURCE_PATH}/fonts DESTINATION ${CURRENT_PACKAGES_DIR}/tools/mapnik)
+    file(COPY ${CURRENT_PACKAGES_DIR}/bin/plugins DESTINATION ${CURRENT_PACKAGES_DIR}/tools/mapnik)
+
+    file(REMOVE ${CURRENT_PACKAGES_DIR}/debug/bin/mapnik-demo ${CURRENT_PACKAGES_DIR}/debug/bin/mapnik-render ${CURRENT_PACKAGES_DIR}/debug/bin/shapeindex)
+
```
Comment (NancyLi1013) [inline ports/mapnik/portfile.cmake:98]: Please remove the extra blank lines.

File: ports/mapnik/portfile.cmake [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+
+endif()
+
+
```
Comment (NancyLi1013) [inline ports/mapnik/portfile.cmake:102]: Please remove the extra blank lines.

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Review by am2222 — COMMENTED** [review]
comments are applied.

**Comment by NancyLi1013** [comment]
Could you please resolve the conflicts again?

**Comment by am2222** [comment]
> Could you please resolve the conflicts again?

Sure, thanks. I am trying to make a set of patches. I will push it soon. I found the reason why it is breaking on windows tests

**Comment by NancyLi1013** [comment]
Hi @am2222 
Could you please resolve the conflicts?

File: ports/mapnik/portfile.cmake [hunk 21]
```diff
@@ -0,0 +1,73 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO mapnik/mapnik
+    REF v3.0.23
+    SHA512 004429defd3047601b63c28bedd57d701ff88cdfb910a9958f4a10a1ac032d4d13316372122aef06eda60f2e4f67600b8842577db455f8c970009a0e86c5c1cf
+    HEAD_REF master
+    PATCHES
+      0001-CMakeFiles.patch 
+      0002-sqlitepixelwidth.patch
```
Comment (NancyLi1013) [inline ports/mapnik/portfile.cmake:9]:
```suggestion
    PATCHES
        0001-CMakeFiles.patch 
        0002-sqlitepixelwidth.patch
```

File: ports/mapnik/portfile.cmake [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+    PATCHES
+      0001-CMakeFiles.patch 
+      0002-sqlitepixelwidth.patch
+)
+
+vcpkg_find_acquire_program(PYTHON2)
+get_filename_component(PYTHON2_DIR "${PYTHON2}" DIRECTORY)
+vcpkg_add_to_path(${PYTHON2_DIR})
+
+set(BOOST_ROOT ${CURRENT_INSTALLED_DIR})
```
Comment (NancyLi1013) [inline ports/mapnik/portfile.cmake:16]: Why do we need to add `set(BOOST_ROOT ${CURRENT_INSTALLED_DIR})` here?

File: ports/mapnik/portfile.cmake [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+    input_postgis  WITH_INPUT_POSTGIS
+    input_raster  WITH_INPUT_RASTER
+    input_shape WITH_INPUT_SHAPE
+    input_sqlite WITH_INPUT_SQLITE
+    input_topojson  WITH_INPUT_TOPOJSON
+    proj4 WITH_PROJ4
+    utils WITH_UTILS
+)
+
+
```
Comment (NancyLi1013) [inline ports/mapnik/portfile.cmake:42]:
```suggestion
```

File: ports/mapnik/portfile.cmake [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+vcpkg_configure_cmake(
+    SOURCE_PATH ${CURRENT_BUILDTREES_DIR}/src/${PORT}
+    PREFER_NINJA # Disable this option if project cannot be built with Ninja
+  OPTIONS   
+      ${FEATURE_OPTIONS}
+    	-DMAPNIK_STATIC_LIB=${MAPNIK_STATIC_LIB}
+    	-DBOOST_PREFIX=${CURRENT_INSTALLED_DIR}/include
+    	-DFREE_TYPE_INCLUDE=${CURRENT_INSTALLED_DIR}/include/freetype2
```
Comment (NancyLi1013) [inline ports/mapnik/portfile.cmake:50]:
```suggestion
    PREFER_NINJA # Disable this option if project cannot be built with Ninja
    OPTIONS   
        ${FEATURE_OPTIONS}
        -DMAPNIK_STATIC_LIB=${MAPNIK_STATIC_LIB}
        -DBOOST_PREFIX=${CURRENT_INSTALLED_DIR}/include
        -DFREE_TYPE_INCLUDE=${CURRENT_INSTALLED_DIR}/include/freetype2
```

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by am2222** [comment]
> Hi @am2222
> Could you please resolve the conflicts?

Hi @NancyLi1013 ,
Thanks for the comments, I will fix them ASAP, I have a deadline by the end of this month and after it I will finish this port, I had some difficulties with making patches and I had to learn how to make them for a package.
Thanks,
Majid

**Comment by NancyLi1013** [comment]
@am2222 
Sorry for the late response.
Please let me know if you need any any from us.

**Comment by am2222** [comment]
> @am2222
> Sorry for the late response.
> Please let me know if you need any any from us.

@NancyLi1013  Hi Nancy,
Thanks so much, Actually I had a question, I wanted to know if there is any tuturials for making patches for vcpkg packages? I am facing errors in making patches and they are mainly because of vcpkg is failing in applying patches.

**Comment by NancyLi1013** [comment]
You can refer to this [doc ](https://github.com/microsoft/vcpkg/blob/master/docs/examples/patching.md).

Hope it can help you some.

Please don't hesitate to ask for help from us if we do can help you.

**Comment by NancyLi1013** [comment]
@am2222 
Could you please resolve the conflicts?

**Comment by am2222** [comment]
> @am2222
> Could you please resolve the conflicts?

Hello @NancyLi1013 
Yes I am working on it. will update it today

File: ports/mapnik/portfile.cmake (hunk 22 above)
Comment (am2222) [reply ports/mapnik/portfile.cmake:30]: It is extra. I am deleting it.

**Comment by NancyLi1013** [comment]
The regressions on x64-windows and x64-windows-static are like this:

```
D:\buildtrees\mapnik\src\mapnik\include\mapnik/geometry/box2d.hpp(115): error C2061: syntax error: identifier 'string'
D:\buildtrees\mapnik\src\mapnik\include\mapnik/geometry/box2d.hpp(132): note: see reference to class template instantiation 'mapnik::box2d<T>' being compiled
D:\buildtrees\mapnik\src\mapnik\include\mapnik/geometry/box2d.hpp(118): error C3646: 'to_string': unknown override specifier
D:\buildtrees\mapnik\src\mapnik\include\mapnik/geometry/box2d.hpp(118): error C2059: syntax error: '('
D:\buildtrees\mapnik\src\mapnik\include\mapnik/geometry/box2d.hpp(118): error C2238: unexpected token(s) preceding ';'
```
Could you please look into this?

**Comment by am2222** [comment]
> The regressions on x64-windows and x64-windows-static are like this:
> 
> ```
> D:\buildtrees\mapnik\src\mapnik\include\mapnik/geometry/box2d.hpp(115): error C2061: syntax error: identifier 'string'
> D:\buildtrees\mapnik\src\mapnik\include\mapnik/geometry/box2d.hpp(132): note: see reference to class template instantiation 'mapnik::box2d<T>' being compiled
> D:\buildtrees\mapnik\src\mapnik\include\mapnik/geometry/box2d.hpp(118): error C3646: 'to_string': unknown override specifier
> D:\buildtrees\mapnik\src\mapnik\include\mapnik/geometry/box2d.hpp(118): error C2059: syntax error: '('
> D:\buildtrees\mapnik\src\mapnik\include\mapnik/geometry/box2d.hpp(118): error C2238: unexpected token(s) preceding ';'
> ```
> 
> Could you please look into this?

@NancyLi1013  Hi,
Yeah sure I will fix it. I was looking for the link of log files to see the errors, but I have lost that link. Thanks

**Comment by am2222** [comment]
@NancyLi1013 Hi,
Sorry for my late reply. Is it possible to share me the full log files? Since I am still trying to find a solution for the errors and could not find any. The even don't happen on my machine with the vcpkg that I have.
Thanks

**Comment by NancyLi1013** [comment]
It seems there are some problems on CI now. I will provide the latest log once it is normal.

**Comment by treigilbert** [comment]
For x64-windows, vcpkg was giving a command parse error for install mapnik[proj4,input_shape]. I had to change all the input_* features to input-* in the CONTROL and portfile.cmake to allow install mapnik[proj4,input-shape].

For x64-windows using Visual Studio 16.7.5 for vcpkg to compile I had to:
Add #include <algorithm> in src\mapnik\include\mapnik\util\char_array_buffer.hpp
Add #include <string> in src\mapnik\include\mapnik\box2d.hpp

For linking errors from including mapnik headers in my code:
In src\mapnik\include\mapnik\enumeration.hpp disable the #ifdef _MCS_VER at line 278. This might be because I was using /Zc:preprocessor when compiling my code.

**Comment by am2222** [comment]
> For x64-windows, vcpkg was giving a command parse error for install mapnik[proj4,input_shape]. I had to change all the input_* features to input-* in the CONTROL and portfile.cmake to allow install mapnik[proj4,input-shape].
> 
> For x64-windows using Visual Studio 16.7.5 for vcpkg to compile I had to:
> Add `#include <algorithm>` in src\mapnik\include\mapnik\util\char_array_buffer.hpp
> Add `#include <string>` in src\mapnik\include\mapnik\box2d.hpp
> 
> For linking errors from including mapnik headers in my code:
> In src\mapnik\include\mapnik\enumeration.hpp disable the #ifdef _MCS_VER at line 278. This might be because I was using /Zc:preprocessor when compiling my code.

Thanks I will apply the changes you mentioned and will push request again. Thanks

**Comment by NancyLi1013** [comment]
@am2222 
Could you please resolve the conflicts?

**Comment by am2222** [comment]
@NancyLi1013 I think so far it does not work on the x64_linux,x64_osx,x86_windows,arm_uwp. I think there is an issue with one of the dependencies. Can I have the logs?

File: scripts/ci.baseline.txt [hunk 25]
```diff
@@ -1887,7 +1887,6 @@ ctp:x64-uwp=skip
 ctp:x64-windows=skip
 ctp:x64-windows-static=skip
 ctp:x86-windows=skip
-protozero:arm-uwp=fail
```
Comment (NancyLi1013) [inline scripts/ci.baseline.txt:1890]: Why do you remove this line here? Seems this port is not related with this PR.

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by NancyLi1013** [comment]
You can get the log here:
https://dev.azure.com/vcpkg/public/_build/results?buildId=42695&view=artifacts&type=publishedArtifacts

File: scripts/ci.baseline.txt (hunk 25 above)
Comment (am2222) [reply scripts/ci.baseline.txt:1890]: I did it by mistake.

**Comment by SNiLD** [comment]
> For x64-windows, vcpkg was giving a command parse error for install mapnik[proj4,input_shape]. I had to change all the input_* features to input-* in the CONTROL and portfile.cmake to allow install mapnik[proj4,input-shape].

These changes are (still) needed. I also needed to add `boost-msm` to the dependencies.

I was building x64-windows with dependencies as static (however gdal, icu, boost-regex, libpq and openssl need to be dynamic).

**Comment by am2222** [comment]
> > For x64-windows, vcpkg was giving a command parse error for install mapnik[proj4,input_shape]. I had to change all the input_* features to input-* in the CONTROL and portfile.cmake to allow install mapnik[proj4,input-shape].
> 
> These changes are (still) needed. I also needed to add `boost-msm` to the dependencies.
> 
> I was building x64-windows with dependencies as static (however gdal, icu, boost-regex, libpq and openssl need to be dynamic).
> 
> After those changes the build was successfull though (haven't linked against it yet).

Are you working with latest version of mapnik? Since I could not make it work with the latest version

**Comment by SNiLD** [comment]
> Are you working with latest version of mapnik? Since I could not make it work with the latest version

I'm using the version you have used in this pull request i.e.:
```
set(GIT_URL "https://github.com/am2222/mapnik-windows.git")
set(GIT_REV "cfa268fa6ab438568b9b5320f6fce4317f0e3d8e")
```

**Comment by am2222** [comment]
@SNiLD  Can you please let me know what other changes did you made that it worked? It works on my local machine but breaks here

**Comment by SNiLD** [comment]
Well first of all I compiled with Visual Studio 2017 Express Edition in x64 mode (from command prompt started with `C:\Program Files (x86)\Microsoft Visual Studio\2017\WDExpress\VC\Auxiliary\Build\vcvarsall.bat x86_x64`).

I used following triplet (named `x64-windows-mapnik`):
```
set(VCPKG_TARGET_ARCHITECTURE x64)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE static)
if(PORT MATCHES "mapnik|gdal|icu|boost-regex|libpq|openssl")
    set(VCPKG_LIBRARY_LINKAGE dynamic)
endif()
```

The build command I used was: `vcpkg install mapnik[input-csv,input-gdal,input-geobuf,input-geojson,input-ogr,input-pgraster,input-postgis,input-raster,input-shape,input-sqlite,input-topojson,proj4] --triplet=x64-windows-mapnik`

I didn't need to do other changes than mentioned in my other post.

**Comment by SNiLD** [comment]
Now that I tried to use mapnik I faced couple of problems with this build. First you need to apply https://github.com/mapbox/variant/pull/182 and https://github.com/mapnik/mapnik/pull/4184 to fix MIN/MAX macro incompatibility with Visual Studio.

Also the dependency headers are not installed by your `CMakeLists.txt`. I can't provide you a pull request because Github [doesn't support](https://stackoverflow.com/questions/6675994/is-it-possible-to-fork-a-fork-in-github) me to have fork of fork when I already have fork of the main project. 💯 
So here is a diff what I needed to add to your CMakeLists.txt:
```patch
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 3e6f4049c..8ae6a09a1 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -163,7 +163,9 @@ set(mapnik_include_paths
     deps/mapbox/polylabel/include
     deps/mapbox/variant/include)
     
-
+install(DIRECTORY deps/agg/include/ DESTINATION include CONFIGURATIONS Release)
+install(DIRECTORY deps/mapbox/geometry/include/ DESTINATION include CONFIGURATIONS Release)
+install(DIRECTORY deps/mapbox/protozero/include/ DESTINATION include CONFIGURATIONS Release)
     
 include_directories(${mapnik_include_paths}  ${thirdparty_include_paths})
 
```

Btw. I have made CMake support for mapnik long time ago based on their SCons scripts (version 3.0.11). Maybe those could help you as well: https://github.com/SNiLD/mapnik/tree/cmake

**Comment by am2222** [comment]
@SNiLD Thanks for the information, I am currently doing my comp exam until mid Dec, I will try to fix them then, 
thanks so much,

**Comment by NancyLi1013** [comment]
@am2222 

Is work still being done for the PR? I noticed that `mapnik ` was also added in another PR #14628.

**Comment by am2222** [comment]
Hi @NancyLi1013 , I could not manage to fix the issues with it yet. I had to take my comp exam and could not spend much time on it. I am not sure if the other PR has solved it or not. If yes we can use it, However I will update the forked mapnik repo with the latest mapnik version

**Comment by mathisloge** [comment]
@am2222 yes, I have resolved the issues. 
There is now a PR @ mapnik for the cmake build https://github.com/mapnik/mapnik/pull/4191

Additionally mapbox/geometry and mapbox/polylabel were added to vcpkg to not to fetch the contents in the port

**Comment by am2222** [comment]
@mathisloge Hi, That is great, so it seems we will finally have it on windows and its nodejs package will also work. that is perfect, probably we need also port its python package as well

**Comment by mathisloge** [comment]
@am2222 that's right. But unfortunately I don't have any spare time to do the node build. 

But it might be easy with https://github.com/cmake-js/cmake-js

**Comment by NancyLi1013** [comment]
@am2222 
Thanks for the PR. I'm closing this for now since there's been no progress for a long time. If you'd like to continue working on it, please reopen and ping us!

**Comment by mathisloge** [comment]
@am2222 if you still using node-mapnik I've added a proof of work https://github.com/mapnik/node-mapnik/pull/976

**Comment by am2222** [comment]
@mathisloge Thanks, are you using the github repo of mapnik for building it? if we can fix this port it will be perfect.

**Comment by mathisloge** [comment]
@am2222 
https://github.com/microsoft/vcpkg/pull/14628 in conjunction with https://github.com/mapnik/mapnik/pull/4191 are working pretty well.

**Comment by mathisloge** [comment]
@am2222 the cmake support is now merged into mapnik@master.

**Comment by am2222** [comment]
@mathisloge thanks man, That is a great new. We finally are able to have windows version alive

**Pull request closed without merging**